src/libsyntax_pos/symbol.rs

```diff
@@ -13,40 +13,64 @@
 //! type, and vice versa.
 
 use hygiene::SyntaxContext;
-use GLOBALS;
+use {Span, DUMMY_SP, GLOBALS};
 
 use serialize::{Decodable, Decoder, Encodable, Encoder};
 use std::collections::HashMap;
 use std::fmt;
+use std::hash::{Hash, Hasher};
 
-#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+#[derive(Copy, Clone, Eq)]
 pub struct Ident {
     pub name: Symbol,
-    pub ctxt: SyntaxContext,
+    pub span: Span,
 }
 
 impl Ident {
+    #[inline]
+    pub const fn new(name: Symbol, span: Span) -> Ident {
+        Ident { name, span }
+    }
+    #[inline]
     pub const fn with_empty_ctxt(name: Symbol) -> Ident {
-        Ident { name: name, ctxt: SyntaxContext::empty() }
+        Ident::new(name, DUMMY_SP)
     }
 
     /// Maps a string to an identifier with an empty syntax context.
     pub fn from_str(string: &str) -> Ident {
         Ident::with_empty_ctxt(Symbol::intern(string))
     }
 
-    pub fn without_first_quote(&self) -> Ident {
-        Ident { name: Symbol::from(self.name.as_str().trim_left_matches('\'')), ctxt: self.ctxt }
+    /// Replace `lo` and `hi` with those from `span`, but keep hygiene context.
+    pub fn with_span_pos(self, span: Span) -> Ident {
+        Ident::new(self.name, span.with_ctxt(self.span.ctxt()))
+    }
+
+    pub fn without_first_quote(self) -> Ident {
+        Ident::new(Symbol::from(self.name.as_str().trim_left_matches('\'')), self.span)
     }
 
     pub fn modern(self) -> Ident {
-        Ident { name: self.name, ctxt: self.ctxt.modern() }
+        Ident::new(self.name, self.span.modern())
+    }
+}
+
+impl PartialEq for Ident {
+    fn eq(&self, rhs: &Self) -> bool {
+        self.name == rhs.name && self.span.ctxt() == rhs.span.ctxt()
+    }
+}
+
+impl Hash for Ident {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.span.ctxt().hash(state);
```

Comment:
For these `Hash` and `PartialEq` I've kept the old behavior - compare and hash idents as `(Name, SyntaxContext)`.

`PartialEq` for `Ident` should either
- not exist (I'd prefer this, but larger structures containing idents want to derive `PartialEq` and `Hash` for whatever reasons), or
- behave like `PartialEq` for `(Name, SyntaxContext)`. `PartialEq` not behaving like this is a notable footgun and source of bugs, we had it couple years ago. Well, and `Hash` should has to be implemented consistently with `PartialEq`.

Reply:
> want to derive `PartialEq` and `Hash` for whatever reasons

What happens if you just remove those? Is `PartialEq` it used only for tests? Could those tests serialize to JSON, use pattern matching, or do something else instead?

Reply:
I'll try.
IIRC, at least `token::Token` contains `Ident` and uses `==` all the time for "simple" variants not actually containing `Ident`s or other data (the case that would be perfectly supported by [`is`](https://github.com/rust-lang/rfcs/issues/929#issuecomment-285602496), btw).

Reply:
Oh, that's a bit of a bummer - and yeah, I agree about `is`.

Reply:
On second thought, should `Token` contain `Ident`, or just `Name` (and keep the `Span` separate)? Or is that too much hassle / not readily supported by existing infrastructure?

Reply:
If lots of spans get moved from the regular HIR nodes to `Ident` and then we don't hash them, that would not be good. But the `HashStable` implementation for `Ident` doesn't ignore the span, right?

Reply:
Yes, `StableHash` doesn't ignore the span (well, after fixing https://github.com/rust-lang/rust/pull/49154#discussion_r175554280).

Also,
>regular HIR nodes

this PR doesn't touch HIR yet, only AST.